src/main/scala/edu/ie3/simona/model/grid/GridModel.scala

```diff
@@ -545,6 +592,132 @@ case object GridModel {
+        measurements,
+        vMax,
+        vMin,
+      )
+  }.toSet
+
+  /** Build a single control group model. Currently, only limit violation
+    * prevention logic is captured: The nodal regulation need is equal to the
+    * voltage change needed to comply with the given thresholds
+    *
+    * @param measurementUnitInput
+    *   Collection of all known [[MeasurementUnitInput]] s
+    * @param measurementConfigs
+    *   Collection of all uuids, denoting which of the [[MeasurementUnitInput]]
+    *   s does belong to this control group
+    * @param vMax
+    *   Upper permissible voltage magnitude
+    * @param vMin
+    *   Lower permissible voltage magnitude
+    * @return
+    *   A [[ControlGroupModel]]
+    */
+  private def buildTransformerControlGroupModel(
+      measurementUnitInput: java.util.Set[MeasurementUnitInput],
+      measurementConfigs: List[String],
+      vMax: Double,
+      vMin: Double,
+  ): ControlGroupModel = {
+    val nodeUuids =
+      determineNodeUuids(measurementUnitInput, measurementConfigs)
+    buildTransformerControlModels(nodeUuids, vMax, vMin)
+  }
+
+  /** Determine the uuids of the nodes to control
+    *
+    * @param measurementUnitInput
+    *   Collection of all known [[MeasurementUnitInput]] s
+    * @param measurementConfigs
+    *   Collection of all uuids, denoting which of the [[MeasurementUnitInput]]
+    *   s does belong to this control group
+    * @return
+    *   A set of relevant nodal uuids
+    */
+  private def determineNodeUuids(
+      measurementUnitInput: java.util.Set[MeasurementUnitInput],
+      measurementConfigs: List[String],
+  ): Set[UUID] = Set.from(
+    measurementUnitInput.asScala
+      .filter(input =>
+        measurementConfigs.contains(input.getUuid.toString) && input.getVMag
+      )
+      .map(_.getNode.getUuid)
+  )
+
+  /** Build a single control group model. Currently, only limit violation
+    * prevention logic is captured: The nodal regulation need is equal to the
+    * voltage change needed to comply with the given thresholds
+    *
+    * @param nodeUuids
+    *   Collection of all relevant node uuids
+    * @param vMax
+    *   Upper permissible voltage magnitude
+    * @param vMin
+    *   Lower permissible voltage magnitude
+    * @return
+    *   A [[ControlGroupModel]]
+    */
+  private def buildTransformerControlModels(
+      nodeUuids: Set[UUID],
+      vMax: Double,
+      vMin: Double,
+  ): ControlGroupModel = {
+    /* Determine the voltage regulation criterion for each of the available nodes */
+    val nodeUuidToRegulationCriterion = nodeUuids.map { uuid =>
+      uuid -> { (complexVoltage: Complex) =>
+        val vMag = complexVoltage.abs
+        vMag match {
+          case mag if mag > vMax =>
+            Some(vMax - mag).map(Each(_))
+          case mag if mag < vMin =>
+            Some(vMin - mag).map(Each(_))
+          case _ => None
+        }
+      }
+    }.toMap
+
+    val harmonizationFunction =
```

Comment:
Same for this function. We'll always have to go back and search for the definition of this function when we start working at the place where it's implemented.